h2o-core/src/main/java/water/util/FrameSizeMonitor.java

```diff
@@ -0,0 +1,151 @@
+            System.out.println("WAIT FOR NEXT " + nextProgress + " " + currentProgress);
+            if (currentProgress >= nextProgress) {
+                if (isMemoryUsageOverLimit(currentProgress, job)) {
+                    job.stop();
+                    break;
+                }
+                if (nextProgress < 0.1f) {
+                    nextProgress *= 2;
+                }
+                System.out.println("NEXT " + nextProgress);
+            }
+            try {
+                Thread.sleep(SLEEP_MS);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+    
+    private boolean isMemoryUsageOverLimit(float progress, Job<Frame> job) {
+        long currentFrameSize = getUsedMemory();
+        long projectedAdditionalFrameSize = (long) ((1 - progress) * (currentFrameSize / progress));
+        long availableMemory = getAvailableMemory();
+        long totalMemory = getTotalMemory();
+        long usableMemory = (long) (availableMemory - (totalMemory * SAFE_FREE_MEM_COEF));
+        if (Log.isLoggingFor(Log.DEBUG)) {
+            Log.debug("FrameSizeMonitor: Frame " + job._result + ": \n" +
+                " used: " + (currentFrameSize / 1024 / 1024) + " MB\n" +
+                " progress: " + (progress) + "\n" +
+                " additional: " + (projectedAdditionalFrameSize / 1024 / 1024) + " MB\n" +
+                " availableMemory: " + (availableMemory / 1024 / 1024) + " MB\n" +
+                " totalMemory: " + (totalMemory / 1024 / 1024) + " MB\n" +
+                " usableMemory: " + (usableMemory / 1024 / 1024) + " MB\n" +
+                " enough: " + (projectedAdditionalFrameSize > usableMemory));
+        }
+        if (projectedAdditionalFrameSize > usableMemory) {
+            Log.err("FrameSizeMonitor: Stopping job " + _jobKey + " writing frame " + job._result +
+                " because the projected size of " + projectedAdditionalFrameSize +
+                " bytes does not safely fit in " + availableMemory + " bytes of available memory.");
+            return true;
+        } else if (Log.isLoggingFor(Log.DEBUG)) {
+            Log.debug("FrameSizeMonitor: Projected memory " + projectedAdditionalFrameSize + " for frame " +
+                job._result + " fits safely into " + availableMemory + " of available memory.");
+        }
+        return false;
+    }
+    
+    private long getUsedMemory() {
+        long usedMemory = 0;
+        synchronized (writers) {
+            for (Map.Entry<FVecParseWriter, Long> e : writers.entrySet()) {
+                FVecParseWriter writer = e.getKey();
+                if (writer.getNvs() != null) {
+                    writers.put(writer, getUsedMemory(writer));
+                }
+                usedMemory += e.getValue();
+            }
+        }
+        return usedMemory;
+    }
+
+    private long getUsedMemory(FVecParseWriter writer) {
```

Comment:
Minor: I can imagine making these methods publicly available in a different class for others to use.

Reply:
yes, but thats out of scope